Python/Product/Analysis/LanguageServer/Server.Completion.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.PythonTools.Interpreter;
+using Microsoft.PythonTools.Parsing;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    public sealed partial class Server {
+        private TaskCompletionSource<bool> _analysisCts;
+
+        public override async Task<CompletionList> Completion(CompletionParams @params) {
+            await _analyzerCreationTask;
+            IfTestWaitForAnalysisComplete();
+
+            var uri = @params.textDocument.uri;
+            // Make sure document is enqueued for processing
+            var openFile = _openFiles.GetDocument(uri);
+
+            _projectFiles.GetAnalysis(@params.textDocument, @params.position, @params._version, out var entry, out var tree);
+            TraceMessage($"Completions in {uri} at {@params.position}");
+
+            tree = GetParseTree(entry, uri, CancellationToken, out _) ?? tree;
+            // Give analysis opportunity to complete
+            if (!entry.IsComplete) {
+                entry.OnNewAnalysis += OnNewAnalysis;
+                _analysisCts = new TaskCompletionSource<bool>();
+                _analysisCts.Task.Wait(300);
```

Comment:
We have `EventTaskSource` in `TestUtilities` which allows you to properly await on events. You can move it down here. See usage at `TestUtilities.Python.EventTaskSources` and `PythonToolsTests.CodeFormatterTests.CodeFormattingTest`.

Reply:
OK

Reply:
If we've never seen a complete analysis for the entry, we should just return a list marked incomplete (which may not even be empty if we've published the in-progress analysis correctly). Better to get an immediate response from the server than to be left hanging.

Reply:
I'm thinking of the `isIncomplete` flag on the resulting list: https://microsoft.github.io/language-server-protocol/specification#textDocument_completion